Modify rule S6466: Extend LaYC content (#3156)
SonarSource/rspec: rules/S6466/java/rule.adoc

An array index out-of-bounds exception is a bug class that occurs in Java when a program tries to access an array element that does not exist.

> This bug can cause your program to crash or behave unexpectedly, and it can also lead to security vulnerabilities.

Author: You are right, its not the same as for C. The comment about security vulnerabilities was already part of the original description and I am also on the fence about it...

Although availability/reliability can be security concerns... I would say I remove the term "security vulnerabilities" here, but keep some of the comments in `Potential Impacts`.

To fix an array index out of bounds exception in Java, you should always ensure that you are accessing array elements within the bounds of the array.

== Why is this an issue?

An array index out-of-bounds exception indicates a bug or a logical error in the
code.

In Java, arrays have a fixed size, and their elements are indexed starting from
`0`, counting up to the last index that is still smaller than the size.
When trying to access an array outside of this range, an
`ArrayIndexOutOfBoundsException` will be thrown and the operation will fail.

include::../impact.adoc[]

== How to fix it

=== Code examples

The following examples contain out-of-bounds accesses to arrays, resulting
in `ArrayIndexOutOfBounds` exceptions.
These situations can be avoided by carefully considering the range of valid
index values, or even better, by comparing indices against the size of a
sequence.

> In this first example, we create an array `arr` containing three elements.
> Since the first element of a list has index `0`, the last valid index is `2`.
> Therefore, an `ArrayIndexOutOfBoundsException` is thrown when accessing `arr` at
> index `3`.

Reviewer: We should avoid 1st person pronouns.
```suggestion
In this first example, the array `arr` contains three elements.
The biggest valid index is `2`.
Therefore, an `ArrayIndexOutOfBoundsException` is thrown when accessing `arr` at
index `3`.
```

==== Noncompliant code example

[source,java,diff-id=1,diff-type=noncompliant]
----
  void fun() {
    int[] arr = {1, 2, 3};

    int x = arr[3]; // Noncompliant: Valid indices are 0, 1 and 2

>     int y = arr[-1]; // Noncompliant: Negative indices are not supported in Java

Reviewer: I would remove the comment about negative indexes, as I don't think thy are a common language feature

Author: They are a feature if you are coming from Python, but I see what you mean. I'll remove it, its enough to mention it in the surrounding text that negative indices are not allowed.

  }
----

==== Compliant solution

[source,java,diff-id=1,diff-type=compliant]
----
  void fun() {
    int[] arr = {1, 2, 3};

    int x = arr[0];
    int y = arr[2];
  }
----

> The following example is similar, but we don't know the concrete length of the
> array `arr` so we use `array.length` instead.
> Still, accessing an array with its size as an index is not correct:

Reviewer: I would simplify 
```suggestion
Accessing an array with its size as an index is never correct:
```
